File: ts/tests/tick.test.ts
import { describe, it, expect } from "vitest";
import { createBootstrap, NoopSigner, Event } from "../src/event.js";
import { Registry, Lifecycle, type Mutation, type Snapshot, type Primitive } from "../src/primitive.js";
import { InMemoryStore } from "../src/store.js";
import { TickEngine } from "../src/tick.js";
import {
  Activation, ActorId, Cadence, ConversationId, EventType, Layer, PrimitiveId, SubscriptionPattern,
} from "../src/types.js";

class CountingPrimitive implements Primitive {
  received = 0;
  private _id: PrimitiveId;
  constructor(name: string, private _layer = 0) { this._id = new PrimitiveId(name); }
  id() { return this._id; }
  layer() { return new Layer(this._layer); }
  cadence() { return new Cadence(1); }
  subscriptions() { return [new SubscriptionPattern("*")]; }
  process(_tick: number, events: Event[], _snap: Snapshot): Mutation[] {
    this.received += events.length;
    return [{ kind: "updateState", primitiveId: this._id, key: "count", value: this.received }];
  }
}

class EmittingPrimitive implements Primitive {
  emissions = 0;
  private _id: PrimitiveId;
  constructor(name: string, private max: number) { this._id = new PrimitiveId(name); }
  id() { return this._id; }
  layer() { return new Layer(0); }
  cadence() { return new Cadence(1); }
  subscriptions() { return [new SubscriptionPattern("*")]; }
  process(_tick: number, events: Event[]): Mutation[] {
    if (events.length === 0 || this.emissions >= this.max) return [];
    this.emissions++;
    return [{ kind: "addEvent", type: new EventType("test.emitted"), source: new ActorId("emitter"), content: { w: this.emissions }, causes: [events[0].id], conversationId: new ConversationId("conv_t") }];
  }
}

function setup(prims: Primitive[] = [], config?: { maxWavesPerTick: number }) {
  const reg = new Registry();
  const store = new InMemoryStore();
  const b = createBootstrap(new ActorId("system"), new NoopSigner());
  store.append(b);
  for (const p of prims) { reg.register(p); reg.activate(p.id()); }
  const engine = new TickEngine(reg, store, config);
  return { reg, store, engine, boot: b };
}

describe("TickEngine", () => {
  it("basic tick", () => {
    const c = new CountingPrimitive("counter");
    const { engine, boot } = setup([c]);
    const r = engine.tick([boot]);
    expect(r.tick).toBe(1);
    expect(c.received).toBe(1);
  });

  it("quiescence", () => {
    const { engine, boot } = setup([new CountingPrimitive("c")]);
    expect(engine.tick([boot]).quiesced).toBe(true);
  });

  it("ripple waves", () => {
    const e = new EmittingPrimitive("e", 3);
    const c = new CountingPrimitive("c");
    const { engine, boot } = setup([e, c]);
    const r = engine.tick([boot]);
    expect(r.waves).toBeGreaterThan(1);
    expect(c.received).toBeGreaterThan(1);
  });

  it("max waves limit", () => {
    const inf: Primitive = {
      id: () => new PrimitiveId("inf"),
      layer: () => new Layer(0),
      cadence: () => new Cadence(1),
      subscriptions: () => [new SubscriptionPattern("*")],
      process: (_t, events) => events.length ? [{ kind: "addEvent", type: new EventType("test.loop"), source: new ActorId("inf"), content: {}, causes: [events[0].id], conversationId: new ConversationId("c") }] : [],
    };
    const { engine, boot } = setup([inf], { maxWavesPerTick: 3 });
    const r = engine.tick([boot]);
    expect(r.waves).toBe(3);
    expect(r.quiesced).toBe(false);
  });

  it("inactive skipped", () => {
    const c = new CountingPrimitive("c");
    const reg = new Registry();
    const store = new InMemoryStore();
    const b = createBootstrap(new ActorId("sys"), new NoopSigner());
    store.append(b);
    reg.register(c); // don't activate
    const engine = new TickEngine(reg, store);
    engine.tick([b]);
    expect(c.received).toBe(0);
  });

  it("tick counter increments", () => {
    const { engine, boot } = setup();
    expect(engine.tick([boot]).tick).toBe(1);
    expect(engine.tick().tick).toBe(2);
  });

  it("layer ordering", () => {
    const order: string[] = [];
    const mkTracker = (name: string, layer: number): Primitive => ({
      id: () => new PrimitiveId(name),
      layer: () => new Layer(layer),
      cadence: () => new Cadence(1),
      subscriptions: () => [new SubscriptionPattern("*")],
      process: () => { order.push(name); return []; },
    });
    const { engine, boot } = setup([mkTracker("high", 5), mkTracker("low", 0), mkTracker("mid", 2)]);
    engine.tick([boot]);
    expect(order).toEqual(["low", "mid", "high"]);
  });

  // --- Layer constraint tests ---

  it("layer constraint blocks when lower layer not invoked", () => {
    const order: string[] = [];
    const mkTracker = (name: string, layer: number): Primitive => ({
      id: () => new PrimitiveId(name),
      layer: () => new Layer(layer),
      cadence: () => new Cadence(1),
      subscriptions: () => [new SubscriptionPattern("*")],
      process: () => { order.push(name); return []; },
    });

    const { engine, boot } = setup([mkTracker("l0", 0), mkTracker("l1", 1)]);

    // Tick 1: Layer 0 runs, Layer 1 blocked
    engine.tick([boot]);
    expect(order).toEqual(["l0"]);

    // Tick 2: Layer 0 stable, Layer 1 now eligible
    order.length = 0;
    engine.tick([boot]);
    expect(order).toContain("l0");
    expect(order).toContain("l1");
    expect(order.indexOf("l0")).toBeLessThan(order.indexOf("l1"));
  });

  it("layer constraint vacuously true for sparse layers", () => {
    let invoked = 0;
    const l1Only: Primitive = {
      id: () => new PrimitiveId("l1_only"),
      layer: () => new Layer(1),
      cadence: () => new Cadence(1),
      subscriptions: () => [new SubscriptionPattern("*")],
      process: () => { invoked++; return []; },
    };

    const { engine, boot } = setup([l1Only]);
    engine.tick([boot]);
    expect(invoked).toBe(1);
  });

  it("layer constraint blocked by dormant lower layer", () => {
    let l1Invoked = 0;

    const l0Dormant: Primitive = {
      id: () => new PrimitiveId("dormant_l0"),
      layer: () => new Layer(0),
      cadence: () => new Cadence(1),
      subscriptions: () => [new SubscriptionPattern("*")],
      process: () => [],
    };

    const l1Active: Primitive = {
      id: () => new PrimitiveId("active_l1"),
      layer: () => new Layer(1),
      cadence: () => new Cadence(1),
      subscriptions: () => [new SubscriptionPattern("*")],
      process: () => { l1Invoked++; return []; },
    };

    const reg = new Registry();
    const store = new InMemoryStore();
    const b = createBootstrap(new ActorId("sys"), new NoopSigner());
    store.append(b);

    reg.register(l0Dormant); // don't activate — stays Dormant
    reg.register(l1Active);
    reg.activate(l1Active.id());

    const engine = new TickEngine(reg, store);
    engine.tick([b]);

    expect(l1Invoked).toBe(0);
  });

  // --- Deferred mutations tests ---

  it("deferred mutations not visible between waves", () => {
    const stateValuesSeen: (unknown | undefined)[] = [];
    let callCount = 0;

    const stateful: Primitive = {
      id: () => new PrimitiveId("stateful"),
      layer: () => new Layer(0),
      cadence: () => new Cadence(1),
      subscriptions: () => [new SubscriptionPattern("*")],
      process: (_tick, events, snap) => {
        callCount++;
        const ps = snap.primitives.get("stateful");
        stateValuesSeen.push(ps?.state?.["count"]);

        if (events.length === 0) return [];

        const result: Mutation[] = [
          { kind: "updateState", primitiveId: new PrimitiveId("stateful"), key: "count", value: callCount },
        ];

        if (callCount === 1) {
          result.push({
            kind: "addEvent",
            type: new EventType("test.ripple"),
            source: new ActorId("stateful"),
            content: {},
            causes: [events[0].id],
            conversationId: new ConversationId("conv_t"),
          });
        }

        return result;
      },
    };

    const { engine, boot } = setup([stateful]);
    const r = engine.tick([boot]);

    expect(r.waves).toBeGreaterThanOrEqual(2);
    // On wave 0, state should be undefined
    expect(stateValuesSeen[0]).toBeUndefined();
    // On wave 1, UpdateState from wave 0 was deferred, so still undefined
    if (stateValuesSeen.length > 1) {
      expect(stateValuesSeen[1]).toBeUndefined();
    }
  });

  it("mixed mutations: AddEvent + UpdateState", () => {
    let emitted = false;
    const pid = new PrimitiveId("mixed");

    const mixed: Primitive = {
      id: () => pid,
      layer: () => new Layer(0),
      cadence: () => new Cadence(1),
      subscriptions: () => [new SubscriptionPattern("*")],
      process: (_tick, events) => {
        if (events.length === 0) return [];

        const result: Mutation[] = [
          { kind: "updateState", primitiveId: pid, key: "processed", value: true },
          { kind: "updateActivation", primitiveId: pid, level: new Activation(0.9) },
        ];

        if (!emitted) {
          emitted = true;
          result.push({
            kind: "addEvent",
            type: new EventType("test.mixed"),
            source: new ActorId("mixed"),
            content: {},
            causes: [events[0].id],
            conversationId: new ConversationId("conv_t"),
          });
        }

        return result;
      },
    };

    const { engine, boot } = setup([mixed]);
    const r = engine.tick([boot]);

    // AddEvent (eager) + UpdateState + UpdateActivation (deferred) = 3+
    expect(r.mutations).toBeGreaterThanOrEqual(3);
  });

  // --- UpdateLifecycle mutation ---

  it("UpdateLifecycle mutation", () => {
    let invoked = 0;
    const pid = new PrimitiveId("updater");

    const updater: Primitive = {
      id: () => pid,
      layer: () => new Layer(0),
      cadence: () => new Cadence(1),
      subscriptions: () => [new SubscriptionPattern("*")],
      process: () => {
        invoked++;
        return [{ kind: "updateLifecycle", primitiveId: pid, state: Lifecycle.Suspending }];
      },
    };

    const { engine, boot } = setup([updater]);

    const r1 = engine.tick([boot]);
    expect(r1.mutations).toBeGreaterThanOrEqual(1);

    // Tick 2: primitive is now Suspending, should NOT be eligible
    const r2 = engine.tick([boot]);
    expect(r2.mutations).toBe(0);
    expect(invoked).toBe(1);
  });

  // --- Subscription filtering ---

  it("subscription filtering: no match", () => {
    const receivedCounts: number[] = [];

    const watcher: Primitive = {
      id: () => new PrimitiveId("watcher"),
      layer: () => new Layer(0),
      cadence: () => new Cadence(1),
      subscriptions: () => [new SubscriptionPattern("trust.*")],
      process: (_t, events) => {
        receivedCounts.push(events.length);
        return [];
      },
    };

    const { engine, boot } = setup([watcher]);
    engine.tick([boot]);
    expect(receivedCounts[0]).toBe(0);
  });

  it("no subscriptions gets no events", () => {
    const receivedCounts: number[] = [];

    const nosubs: Primitive = {
      id: () => new PrimitiveId("nosubs"),
      layer: () => new Layer(0),
      cadence: () => new Cadence(1),
      subscriptions: () => [],
      process: (_t, events) => {
        receivedCounts.push(events.length);
        return [];
      },
    };

    const { engine, boot } = setup([nosubs]);
    engine.tick([boot]);
    expect(receivedCounts[0]).toBe(0);
  });

  // --- Wave limit ---

  it("wave limit with custom config", () => {
    const inf: Primitive = {
      id: () => new PrimitiveId("always"),
      layer: () => new Layer(0),
      cadence: () => new Cadence(1),
      subscriptions: () => [new SubscriptionPattern("*")],
      process: (_t, events) =>
        events.length
          ? [{ kind: "addEvent", type: new EventType("test.always"), source: new ActorId("a"), content: {}, causes: [events[0].id], conversationId: new ConversationId("c") }]
          : [],
    };

    const { engine, boot } = setup([inf], { maxWavesPerTick: 5 });
    const r = engine.tick([boot]);
    expect(r.waves).toBe(5);
    expect(r.quiesced).toBe(false);
  });
});

File: rust/tests/tick_test.rs
use std::collections::BTreeMap;
use serde_json::Value;
use eventgraph::event::*;
use eventgraph::primitive::*;
use eventgraph::store::InMemoryStore;
use eventgraph::tick::*;
use eventgraph::types::*;

struct CounterPrimitive {
    pid: PrimitiveId,
}

impl CounterPrimitive {
    fn new(name: &str) -> Self {
        Self { pid: PrimitiveId::new(name).unwrap() }
    }
}

impl Primitive for CounterPrimitive {
    fn id(&self) -> PrimitiveId { self.pid.clone() }
    fn layer(&self) -> Layer { Layer::new(0).unwrap() }

    fn process(&self, _tick: u64, events: &[Event], _snapshot: &Snapshot) -> Vec<Mutation> {
        vec![Mutation::UpdateState {
            primitive_id: self.pid.clone(),
            key: "count".to_string(),
            value: Value::Number(events.len().into()),
        }]
    }

    fn subscriptions(&self) -> Vec<SubscriptionPattern> {
        vec![SubscriptionPattern::new("*").unwrap()]
    }

    fn cadence(&self) -> Cadence { Cadence::new(1).unwrap() }
}

struct EmitterPrimitive {
    pid: PrimitiveId,
}

impl EmitterPrimitive {
    fn new(name: &str) -> Self {
        Self { pid: PrimitiveId::new(name).unwrap() }
    }
}

impl Primitive for EmitterPrimitive {
    fn id(&self) -> PrimitiveId { self.pid.clone() }
    fn layer(&self) -> Layer { Layer::new(0).unwrap() }

    fn process(&self, _tick: u64, events: &[Event], snapshot: &Snapshot) -> Vec<Mutation> {
        let count = snapshot.primitives.get(self.pid.value())
            .and_then(|s| s.state.get("emitted"))
            .and_then(|v| v.as_u64())
            .unwrap_or(0);

        if count > 0 { return vec![]; }

        if events.is_empty() { return vec![]; }
        let cause = events[0].id.clone();

        let mut content = BTreeMap::new();
        content.insert("source".to_string(), Value::String("emitter".to_string()));

        vec![
            Mutation::UpdateState {
                primitive_id: self.pid.clone(),
                key: "emitted".to_string(),
                value: Value::Number(1.into()),
            },
            Mutation::AddEvent {
                event_type: EventType::new("test.emitted").unwrap(),
                source: ActorId::new("emitter").unwrap(),
                content,
                causes: vec![cause],
                conversation_id: ConversationId::new("conv_test").unwrap(),
            },
        ]
    }

    fn subscriptions(&self) -> Vec<SubscriptionPattern> {
        vec![SubscriptionPattern::new("*").unwrap()]
    }

    fn cadence(&self) -> Cadence { Cadence::new(1).unwrap() }
}

fn setup_engine(prims: Vec<Box<dyn Primitive>>) -> TickEngine {
    let mut registry = Registry::new();
    for p in prims {
        let pid = p.id();
        registry.register(p).unwrap();
        registry.activate(&pid).unwrap();
    }
    let store = InMemoryStore::new();
    TickEngine::new(registry, store, None)
}

#[test]
fn tick_increments() {
    let mut engine = setup_engine(vec![Box::new(CounterPrimitive::new("counter"))]);
    let boot = create_bootstrap(ActorId::new("alice").unwrap(), &NoopSigner, 1);
    let r = engine.tick(Some(vec![boot]));
    assert_eq!(r.tick, 1);
    assert!(r.mutations > 0);
}

#[test]
fn tick_quiesces_with_no_events() {
    let mut engine = setup_engine(vec![Box::new(CounterPrimitive::new("counter"))]);
    let boot = create_bootstrap(ActorId::new("alice").unwrap(), &NoopSigner, 1);
    let r = engine.tick(Some(vec![boot]));
    // Wave 1 processes, wave 2 finds no new events -> quiesces
    assert!(r.quiesced);
}

#[test]
fn emitter_causes_ripple() {
    let mut engine = setup_engine(vec![
        Box::new(EmitterPrimitive::new("emitter")),
        Box::new(CounterPrimitive::new("counter")),
    ]);
    let boot = create_bootstrap(ActorId::new("alice").unwrap(), &NoopSigner, 1);
    let r = engine.tick(Some(vec![boot]));
    // Emitter emits in wave 1, counter processes in wave 2
    assert!(r.waves >= 2);
}

#[test]
fn empty_tick() {
    let mut engine = setup_engine(vec![Box::new(CounterPrimitive::new("counter"))]);
    let r = engine.tick(None);
    // Wave 1 with no events, counter still processes (wave > 0 check), mutations = 1
    assert_eq!(r.tick, 1);
}

#[test]
fn inactive_primitives_skipped() {
    let mut registry = Registry::new();
    let p = Box::new(CounterPrimitive::new("counter"));
    registry.register(p).unwrap();
    // Don't activate - stays Dormant
    let store = InMemoryStore::new();
    let mut engine = TickEngine::new(registry, store, None);

    let boot = create_bootstrap(ActorId::new("alice").unwrap(), &NoopSigner, 1);
    let r = engine.tick(Some(vec![boot]));
    assert_eq!(r.mutations, 0);
}

#[test]
fn multiple_ticks() {
    let mut engine = setup_engine(vec![Box::new(CounterPrimitive::new("counter"))]);
    let boot = create_bootstrap(ActorId::new("alice").unwrap(), &NoopSigner, 1);
    engine.tick(Some(vec![boot]));
    let r2 = engine.tick(None);
    assert_eq!(r2.tick, 2);
}

#[test]
fn cadence_respected() {
    struct SlowPrimitive { pid: PrimitiveId }
    impl Primitive for SlowPrimitive {
        fn id(&self) -> PrimitiveId { self.pid.clone() }
        fn layer(&self) -> Layer { Layer::new(0).unwrap() }
        fn process(&self, _tick: u64, _events: &[Event], _snapshot: &Snapshot) -> Vec<Mutation> {
            vec![Mutation::UpdateState {
                primitive_id: self.pid.clone(),
                key: "ran".to_string(),
                value: Value::Bool(true),
            }]
        }
        fn subscriptions(&self) -> Vec<SubscriptionPattern> {
            vec![SubscriptionPattern::new("*").unwrap()]
        }
        fn cadence(&self) -> Cadence { Cadence::new(3).unwrap() }
    }

    let mut engine = setup_engine(vec![Box::new(SlowPrimitive {
        pid: PrimitiveId::new("slow").unwrap(),
    })]);

    let boot = create_bootstrap(ActorId::new("alice").unwrap(), &NoopSigner, 1);
    let r1 = engine.tick(Some(vec![boot]));
    // tick 1: last=0, 1-0=1 < 3 → skip (cadence not met)
    assert_eq!(r1.mutations, 0);

    let _r2 = engine.tick(None); // tick 2: 2-0=2 < 3 → skip
    assert_eq!(_r2.mutations, 0);

    let r3 = engine.tick(None); // tick 3: 3-0=3 >= 3 → runs
    assert!(r3.mutations > 0);
}
